Hook Estimator: Node Setup and Publishing Functionality

0Dema0 opened the pull request:
Please review the implementation of hook_estimator.py.

This PR adds the node setup and publishing logic for the hook estimator.
The node should correctly initialize and publish the expected data.
Note: The rope is currently assumed to be red (this behavior remains unchanged).

Feedback on functionality and code style is welcome!

KKalem commented on external_packages/smarc_modelling:
do not commit changes to submodules. you must undo this change.

KKalem commented on perception/alars/auv_detector/auv_detector/hook_estimator.py:
```diff
@@ -0,0 +1,1020 @@
+from std_msgs.msg import Float32MultiArray, Header
+from geometry_msgs.msg import PoseStamped, AccelStamped, Vector3Stamped
+from sensor_msgs.msg import Imu
+from drone_msgs.msg import Topics as DroneTopics
```
drone_msgs are deprecated, use dji_msgs instead.

KKalem commented on perception/alars/auv_detector/auv_detector/hook_estimator.py:
```diff
@@ -0,0 +1,1020 @@
+
+        self.subscription_camera = self.create_subscription(
+            Image,
+            '/Quadrotor/core/fpcamera/image',
```
use the Topics.msg system. Hardcoded topics are not allowed.
The first part `/Quadrotor` is the name of the robot, and your node should work for _any robot name_ so it will be namespaced by ros. None of your topics should start with `/` and none should have the robot's name in them.

For example, this topic parameter should be `DjiTopics.CAMERA_TOPIC` after importing `from dji_msgs.msg import Topics as DjiTopics`. Take this as an example for all things related to topics: https://github.com/smarc-project/smarc2/blob/humble/vehicles/hardware/dji/dji_captain/dji_captain/dji_captain.py

KKalem commented on perception/alars/auv_detector/auv_detector/hook_estimator.py:
```diff
@@ -0,0 +1,1020 @@
+
+        self.subscription_accel = self.create_subscription(
+            Imu,
+            '/Quadrotor/core/imu',
```
use smarc_msgs.msg.ODOM_TOPIC instead. IMU is too noisy for direct use.

KKalem commented on perception/alars/auv_detector/auv_detector/hook_estimator.py:
```diff
@@ -0,0 +1,1020 @@
+        self.tf_buffer = Buffer()
+        self.tf_listener = TransformListener(self.tf_buffer, self)
+
+        # Set frames
```
_ALL_ of these frames should be coming from dji_msgs.msg.LINKS similar to topics. 
however, links can not be namespaced like topics, so your node needs to take `robot_name` as a rosparam (and the launch file that launches it).

example: https://github.com/smarc-project/smarc2/blob/d07b9ffea6c5b28e17e5665fcecc90a6250abca3/vehicles/hardware/dji/dji_captain/dji_captain/dji_captain.py#L106

KKalem commented on perception/alars/auv_detector/auv_detector/hook_estimator.py:
```diff
@@ -0,0 +1,1020 @@
+
+        self.static_tf_timer = self.create_timer(2.0, self.init_static_transfroms_once)
+
+        self.est_hook_pub = self.create_publisher(PointStamped, '/hook_est_position', 10)
```
add your new topics to dji_msgs topics. follow the naming scheme in the file. include "ESTIMATE" for things you estimate :)

KKalem commented on perception/alars/auv_detector/auv_detector/hook_estimator.py:
```diff
@@ -0,0 +1,1020 @@
+        ]
+
+        def full_frame(name):
+            return f'Quadrotor/{name}'
```
robot name needs to be a parameter

KKalem commented:
Missing features:
- enable_hook_estimator, disable_hook_estimator services that stop this from working. type: trigger. service names also need to go into dji_msgs.msg topics file.
- a parameter to disable all on-screen plotting and drawing and such. this will run on a computer with no screens attached :)

0Dema0 commented:
I should have changed the topics and frames correctly. Also, I added both the missing features, but I am unsure about the trigger (I basically added it as a parameter in the launch file).

KKalem commented on messages/dji_msgs/msg/Topics.msg:
```diff
@@ -1,9 +1,30 @@
+string LOAD_CELL_WEIGHT_TOPIC = 'sensor/load_cell_weight' # Float32
+
+# Used by hook_estimator
+string CAMERA_TOPIC = 'core/fpcamera/image'
```
why not use existing gimbal camera raw topic?

KKalem commented on messages/dji_msgs/msg/Topics.msg:
```diff
@@ -1,9 +1,30 @@
+# Used by hook_estimator
+string CAMERA_TOPIC = 'core/fpcamera/image'
+string HOOK_ESTIMATE_TOPIC = 'hook_estimate'
+string HOOK_EKF_ESTIMATE_TOPIC = 'hook_ekf_estimate'
```
please match the other topics
`ESTIMATED_HOOK_TOPIC = 'alars_detection/hook' #PoseStamped` 
(PoseStamped is the type you'll be publishing, might be sth else)

KKalem commented on messages/dji_msgs/msg/Links.msg:
```diff
@@ -11,4 +11,12 @@ string HOME_POINT = 'home_point'
+
+
+# Used by hook_estimator
+string CAMERA_LINK = 'camera_link'
```
use gimbal camera link

KKalem commented on messages/dji_msgs/msg/Links.msg:
```diff
@@ -11,4 +11,12 @@ string HOME_POINT = 'home_point'
+# Used by hook_estimator
+string CAMERA_LINK = 'camera_link'
+string WINCH_LINK = 'winch_link'
+string IMU_LINK = 'imu_link'
```
why do you need to know _where_ the imu is? You should be using the fused estimate like here: https://github.com/smarc-project/smarc2/blob/3e0210ba74c582b49b2b2c03aef93c0e85346df9/behaviours/alars/alars/alars_recover_action_server.py#L383

0Dema0 replied:
I don't, it was something I used in the US, I just deleted it

KKalem commented on messages/dji_msgs/msg/Links.msg:
```diff
@@ -11,4 +11,12 @@ string HOME_POINT = 'home_point'
+
+# Used by hook_estimator
+string CAMERA_LINK = 'camera_link'
+string WINCH_LINK = 'winch_link'
```
This must be published by something. Like the captain or a static transform publisher. So far I've been adding static transforms to the captain for convenience. Can you do that? Follow the example of the gimbal_camera. You just need to change the position from all 0s. In general, whatever captain doesn't produce, is not going to exist in real tests.

0Dema0 replied:
I added the winch to the captain. Not the hook, the entire point of estimating its position is that we won't have it. I put some values for translation and rotation, of course we should change them according to the real thing.

KKalem commented on messages/dji_msgs/msg/Topics.msg:
```diff
@@ -1,9 +1,30 @@
-string CONTROLLER_VIBRATION_TOPIC = 'joy/set_feedback'
-string TTS_TOPIC = 'speak'
 
 
```
y u delete my comment bruv

KKalem commented on perception/alars/auv_detector/auv_detector/hook_estimator.py:
```diff
@@ -29,30 +32,35 @@
-            self.imu_callback,
+        self.subscription_odom = self.create_subscription(
+            Odometry,
+            'smarc/odom',
```
hardcoded topic. you already imported the smarc msgs, use it yo

KKalem commented on perception/alars/auv_detector/setup.py:
```diff
@@ -25,7 +25,8 @@
             'hook_estimator = auv_detector.hook_estimator:main',
-	        'estimator = auv_detector.auv_detector:main', 
+            'auv_buoy_detector = auv_detector.auv_buoy_detector:main', 
+	          'estimator = auv_detector.auv_detector:main', 
```
extra space :)

KKalem commented on messages/dji_msgs/msg/Links.msg:
```diff
@@ -11,4 +11,12 @@ string HOME_POINT = 'home_point'
+string CAMERA_LINK = 'camera_link'
+string WINCH_LINK = 'winch_link'
+string IMU_LINK = 'imu_link'
+string HOOK = 'Hook'
```
convention is have "link" in tf-related items. plz.

KKalem commented on perception/alars/auv_detector/auv_detector/hook_estimator.py:
```diff
@@ -91,30 +99,41 @@ def __init__(self):
+
+        # Create enable/disable services
+        self.enable_service = self.create_service(Trigger, 'enable_hook_estimator', self.enable_callback)
+        self.disable_service = self.create_service(Trigger, 'disable_hook_estimator', self.disable_callback)
```
add these service topics to the dji topics.msg like the other ones there.

KKalem commented:
> I should have changed the topics and frames correctly. Also, I added both the missing features, but I am unsure about the trigger (I basically added it as a parameter in the launch file).

Yep, you almost got it. Few changes to make still tho.

Trigger looks good, just gotta add the service topic to the msg file like everything else.

0Dema0 commented:
Everything should be fine, apart from the following:
- I guess I should use the hook_link only during simulations (hence I should add some check on that, or just add the hook in the captain as a dummy)
- Rope color and camera intrinsics are still the same, before usage one should set them correctly (I also wrote this in the readme file)

KKalem commented:
See my slack message for review.
Marked draft. Mark ready for review when you need me to look at it again.

0Dema0 committed:
re-cloned repository with correct changes added

0Dema0 commented:
Now it builds. My last comment is still valid (hook_link checks and rope color\camera intrinsics), but I removed the prints, added the visualization as a rosparam and set sim_time

0Dema0 committed:
added hook estimator scripts and restored diff

0Dema0 commented:
Should be good now, the scripts were "untracked", and I restored the diff

0Dema0 committed:
set rope length as a rosparam
